### ISSUE TITLE
Sync too slow

### ISSUE THREAD

**Issue by shadowpigy** [eqlabs/pathfinder#1376]
Starting the process with debug mode, there is no errors, but the block time is too long, leading to around 300 blocks behind global height.
```
2023-09-15T10:20:08 DEBUG Updated sync status status=starting: (0x045E8B53D12E866C9E27A118A072EFFC01A428E64F18C0740331AFB7BB8AB9F4, 224377), current: (0x028A360406F306158693BEDA20584ECD34E0C1092DF4B7422234CA2D0F34D5A0, 224383), highest: (0x064466D062403FEB27E7D6FC7D33B870D3CAE57D9AF0368B8DF324753C9EBE28, 225247)
2023-09-15T10:20:13 DEBUG Updated Starknet state with block 224384 after 24.817001s (8.016595s avg). contracts (0.000160825s), 678 storage updates (24.816929s). Block downloaded in 0.80516016s, state diff in 0.037705757s
2023-09-15T10:20:24 DEBUG Updated sync status status=starting: (0x045E8B53D12E866C9E27A118A072EFFC01A428E64F18C0740331AFB7BB8AB9F4, 224377), current: (0x050C09227D83382376591C0872AEAEA92C42F11F502DF7E3084AEA1C67CD76A5, 224384), highest: (0x01DAA71742BD69C07A06933725DD47F905D4F784579BB31DA715AAC5D608240C, 225248)
.............
2023-09-15T10:20:41 DEBUG Updated Starknet state with block 224385 after 27.270073s (8.979269s avg). contracts (0.000156577s), 746 storage updates (27.270004s). Block downloaded in 0.81385016s, state diff in 0.17424121s
2023-09-15T10:20:57 DEBUG Updated sync status status=starting: (0x045E8B53D12E866C9E27A118A072EFFC01A428E64F18C0740331AFB7BB8AB9F4, 224377), current: (0x047BECE7ABA62675B6DCEFE9F35E90DB8E9E1583D1541D9F9A0CA9A9ADE9DDD5, 224385), highest: (0x01F17D048B4BF97937DA8A6100C01B6F515F480D8A077590218C1B901444118E, 225250)
```

**Comment by Mirko-von-Leipzig** [comment]
Apologies for the delayed response. We've actually been looking into this exact issue.

## TL;DR

Please try our pruned mainnet snapshot ([link](https://pub-1fac64c3c0334cda85b45bcc02635c32.r2.dev/mainnet-241820-pruned-until-238777.sqlite.zst)). This should help reduce the times.

SHA256 sum is `f0a2b1274a396a77f0601fbff0e8828ba91949540bd7dcd627c6e73018326f2e` for the compressed file.

We've experienced some unreliability with the link above -- downloads over HTTPS are prone to failures. We recommend using [rclone](https://rclone.org/) for reliable downloads. After installing rclone you have to add the following section to $HOME/.config/rclone/rclone.conf:
```
[pathfinder-snapshots]
type = s3
provider = Cloudflare
endpoint = https://cbf011119e7864a873158d83f3304e27.r2.cloudflarestorage.com/
```
Then you can download the file using the following command:
rclone copy -P pathfinder-snapshots:pathfinder-snapshots/mainnet-241820-pruned-until-238777.sqlite.zst .

## Longer explanation

Mainnet state has grown dramatically in the past few months causing specifically the state update portion of our sync to no longer keep up. Our temporary solution for this is to prune the state -- which is what is provided in the database snapshot linked above.

This has shown improvements of between 4x to 2x.

More specifically, disks with a high latency or low IOPS (such as network disks commonly employed in cloud instances) suffer from this problem. Updating the state trees requires many small, random reads and writes. Using a locally attached SSD has update times of 1-3s whereas these network style disks show anything from 20-50s times currently. Using the pruned snapshot brings this down to a more reasonable 5-10s. Still not great but at least useable.

## Future solutions

We're investigating several options as fixes within the node client itself. Some avenues:

- Automatic pruning so that the state doesn't keep growing endlessly. However since this would require more reads and writes again, this might be counter-productive in the short term.
- Different database schemas. This may reduce lookup times but the reads and writes remain random and small so its unclear how much this will help in practice. None-the-less we are trying these out.
- Chunking / batching reads and writes. As before, the random nature means its unclear whether this will help much in the long run.
- Only persist every every N blocks. This will require quite a few changes and will take a while to implement and test.

**Comment by shadowpigy** [comment]
Hi, regard to the snapshot you provided, do you update the snapshot regularly or it is just for single usage?

**Comment by Mirko-von-Leipzig** [comment]
We do generally host snapshots which are updated semi-regularly only. Its something we need to properly setup ito operational management.

However, this "pruned" snapshot is hopefully a once-off. We want to have a better solution for pruning state.

**Comment by shadowpigy** [comment]
Yeah. Hopefully. btw is there any pruning configs in the client we could set ourselves?

**Comment by Mirko-von-Leipzig** [comment]
Not yet, currently its a manual tool we made. Once we implement pruning within the client we will make sure its configurable.
